posts/nyc-subway-data-with-kafka-streams-and-jhipster-part-2.md

```diff
@@ -0,0 +1,94 @@
+
+```java
+@StreamListener(MessagingConfiguration.MtaStream.INPUT)
+public void saveSubwayCount(SubwayCount subwayCount) {
+    // Save measurement in influxdb
+    influxDB.write(Point.measurement("line")
+        .time(subwayCount.getStart().toEpochMilli(), TimeUnit.MILLISECONDS)
+        .tag("route", subwayCount.getRoute())
+        .addField("count", subwayCount.getCount())
+        .build());
+}
+```
+
+The method above get triggered when a message is published in the topic `mta-stream` and will then write the message in InfluxDB. Spring Cloud Stream makes things simple by converting the Kafka message from a JSON format to the custom `SubwayCount` class.
+
+InfluxDB's measurement is conceptually similar to a table and all messages are inserted in the `line` measurement. The subway's route is used as a tag to allow filtering when it will be used in Grafana. There is only one field for the point and it is the number of active subways. Finally the time of the point is the start timestamp of the Kafka's window.
+
+# Grafana configuration and dashboard creation
+
+Once again, thanks to Docker for making the Grafana setup easy:
+
+```docker run -d -p 3000:3000 grafana/grafana```
+
+A data source for InfluxDB must be created in order to build a dashboard, here is a screenshot of my configuration:
+
+![mta-kafka-datasource](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/11/mta-kafka-datasource.png)
+
+The ip must be the one of the InfluxDB's container, it can be retrieved using the command below:
+
+```docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' container_name_or_id```
+
+Any dashboard can be created with the datasource, I decided to have one with a panel that will display the number of subways per line.
+
+![mta-kafka-2018-11-05](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/11/mta-kafka-2018-11-05.png)
+
+Lines are grouped to avoid having too many series and the value is the average number of running trains for a 30mins interval. The above screenshot shows the busiest lines on Monday November 5 in 2018 which is a regular work day.
+
+This [JSON configuration](https://raw.githubusercontent.com/Falydoor/mta-kafka/master/dashboard.json) contains a dashboard with the panel from the screenshot and it can be easily imported in a different Grafana instance.
+
+# Conclusion
+
+From the screenshot, we can see that the two times where the number of active subways reach its peak are 9am and 5pm. The time of the day where the number is the lowest is between midnight and 5am. The red line (or 1-2-3) is the one with the most running trains in average during a day which makes sense since it is one of the longest line.
```

Comment:
```suggestion
From the screenshot, we can see that the two times where the number of active subways reach its peak are 9am and 5pm. The time of the day where the number is the lowest is between midnight and 5am. The red line (or 1-2-3) is the one with the most running trains on average during a day which makes sense since it is one of the longest lines.
```